XiEditor/ShadowView.swift

```diff
@@ -0,0 +1,99 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Cocoa
+
+class ShadowView: NSView, CALayerDelegate {
+
+    override var wantsUpdateLayer: Bool {
+        return true
+    }
+
+    fileprivate var leftShadow = CAGradientLayer()
+    fileprivate var rightShadow = CAGradientLayer()
+
+    /// The x coordinate of the leftmost edge of the left shadow.
+    /// If there is a gutter, this should equal the gutter's right edge.
+    var leftShadowMinX: CGFloat = 0 {
+        didSet {
+            self.needsDisplay = true
+        }
+    }
+
+    /// If set to true, this view will display a shadow at the left side of the view.
+    var showLeftShadow: Bool = true {
+        didSet {
+            if showLeftShadow != oldValue {
+                updateVisibility()
+            }
+        }
+    }
+
+    /// If set to true, this view will display a shadow at the right side of the view.
+    var showRightShadow: Bool = true {
+        didSet {
+            if showRightShadow != oldValue {
+                updateVisibility()
+            }
+        }
+    }
+
+    fileprivate var shadowColor = NSColor.shadowColor
+
+    func updateShadowColor(newColor: NSColor?) {
+        if newColor ?? NSColor.shadowColor != self.shadowColor {
+            self.shadowColor = newColor ?? NSColor.shadowColor
+            self.needsDisplay = true
+        }
+    }
+
+    fileprivate func updateVisibility() {
+        self.isHidden = !(showLeftShadow || showRightShadow)
+        leftShadow.isHidden = !showLeftShadow
+        rightShadow.isHidden = !showRightShadow
+    }
+
+    func setup() {
+        self.layerContentsRedrawPolicy = .onSetNeedsDisplay
+        self.leftShadow.delegate = self
+        self.rightShadow.delegate = self
+        self.layer!.addSublayer(leftShadow)
+        self.layer!.addSublayer(rightShadow)
+
+        updateVisibility()
+    }
+
+    override func updateLayer() {
+        leftShadow.frame = CGRect(x: leftShadowMinX, y: self.bounds.origin.y,
+                                  width: 4, height: self.bounds.height)
+
+        rightShadow.frame = CGRect(x: self.bounds.width - 4, y: self.bounds.origin.y,
+                                   width: 4, height: self.bounds.height)
+
+        leftShadow.colors = [shadowColor.cgColor, NSColor.clear.cgColor]
+        leftShadow.transform = CATransform3DMakeRotation((3 * CGFloat.pi) / 2, 0, 0, 1)
+        leftShadow.opacity = 0.4
+        leftShadow.autoresizingMask = .layerHeightSizable
+
+        rightShadow.colors = [shadowColor.cgColor, NSColor.clear.cgColor]
+        rightShadow.transform = CATransform3DMakeRotation(CGFloat.pi / 2, 0, 0, 1)
+        rightShadow.opacity = 0.4
+        rightShadow.autoresizingMask = [.layerHeightSizable, .layerMinXMargin]
+    }
+
+    // by default hiding and revealing our shadows is animated, which is distracting
+    func action(for layer: CALayer, forKey event: String) -> CAAction? {
```

Comment:
I personally found that it looked really nice. Maybe the duration could be reduced?

Reply:
I wouldn't be opposed to a very short-duration animation, as long as there was some way to specify that the animation should begin from whatever the existing animation state is, so that we don't have animations that interrupt each other; I looked into this briefly but it seems harder to do in AppKit than it would've been in UIKit.

When animations were on, it also caused the right shadow view to animate when you first opened the window, which would have to be fixed.

I initially thought no animations would be a bit jarring, but playing around with it I think it feels okay; and looking at other editors (particularly sublime) this is their behaviour as well.